### ISSUE TITLE
Math.SinCos can be slower than Sin + Cos in some scenarios.

### ISSUE THREAD

**Issue by tannergooding** [dotnet/runtime#48776]
The codegen for `Math.SinCos` can currently be problematic on GCC and MSVC leading to decreased perf.

https://godbolt.org/z/K5dad3 shows the codegen for:
```cpp
#include <math.h>

#ifdef _MSC_VER
#pragma float_control(precise, off)
#endif

void SinCos1(float x, float* pSin, float* pCos)
{
#ifdef _MSC_VER
    *pSin = sinf(x);
    *pCos = cosf(x);
#else
    sincosf(x, pSin, pCos);
#endif
}
```

In particular, GCC and MSVC end up with many shuffles between registers and to/from the stack causing a noticeable perf slowdown across hundreds of calls.

https://godbolt.org/z/sG6dTW shows comparison of several other possible variants, but none is "ideal" across all 3 compilers. We should investigate this further to determine what can be done to ensure that `SinCos` is at least as fast as calling `Sin` and `Cos` independently, possibly following up with the MSVC team to improve the codegen here.

**Comment by tannergooding** [comment]
This was reported by a customer in the .NET 6 Preview 1 blogpost here: https://devblogs.microsoft.com/dotnet/announcing-net-6-preview-1/comment-page-6/#comment-8548

**Comment by echesakovMSFT** [comment]
I was thinking how to mitigate the issue and one of the ideas was to have two separate calls to `Math.Sin` and `Math.Cos` (under `TARGET_WINDOWS`) in managed implementation of `Math.SinCos` in S.P.CoreLib to avoid the issue with MSVC codegen. 

Unfortunately, such approach doesn't improve the situation much. 

As an example, this is what `Math.SinCos` would have looked if we had that implementation instead of a native call to `SinCos(double, double*, double*)`

```asm
G_M57534_IG01:              ;; offset=0000H
       56                   push     rsi
       4883EC30             sub      rsp, 48
       C5F877               vzeroupper 
       488BF1               mov      rsi, rcx
       C5F828C1             vmovaps  xmm0, xmm1
						;; bbWeight=1    PerfScore 2.75
G_M57534_IG02:              ;; offset=000FH
       C5FB11442448         vmovsd   qword ptr [rsp+48H], xmm0
       E8E6CBB35F           call     System.Math:Sin(double):double
       C5FB11442428         vmovsd   qword ptr [rsp+28H], xmm0
       C5FB10442448         vmovsd   xmm0, qword ptr [rsp+48H]
       E815C0B35F           call     System.Math:Cos(double):double
       C5FB104C2428         vmovsd   xmm1, qword ptr [rsp+28H]
       C5FB110E             vmovsd   qword ptr [rsi], xmm1
       C5FB114608           vmovsd   qword ptr [rsi+8], xmm0
       488BC6               mov      rax, rsi
						;; bbWeight=1    PerfScore 8.25
G_M57534_IG03:              ;; offset=003DH
       4883C430             add      rsp, 48
       5E                   pop      rsi
       C3                   ret      
```

**Comment by tannergooding** [comment]
> I was thinking how to mitigate the issue and one of the ideas was to have two separate calls to Math.Sin and Math.Cos (under TARGET_WINDOWS) in managed implementation of Math.SinCos in S.P.CoreLib to avoid the issue with MSVC codegen.

The two calls in C++ get intrinsified to a call to the internal `__sincos` helper method which does "less work". I think the issue is really due to the shuffling that is happening since C++ has to assume its being used a particular way that doesn't match up with the intrinsic signature.

**Comment by echesakovMSFT** [comment]
Yes, I understood this part. I was more thinking what could be done now (in 6.0.0) to mitigate that problem.

@tannergooding Do you know if the helper can be called directly? It looks that it has some "custom ABI" - the values of both `sin` and `cos` returned as components of `xmm0` register. As an idea, we could introduce a "private" hardware intrinsic on win-x64 `Vector128<double> SinCos(double)` (that would unfold into a call to the helper), then we could use the intrinsic in the managed implementation of `Math.SinCos`. Basically, moving the unpacking of the values from `xmm0` into managed code.

**Comment by echesakovMSFT** [comment]
We can tweak the native helper signature - only return a value of `cos(x)` by reference.

```c++
float SinCos(float x, float* pCos)
{
    *pCos = cosf(x);
    return sinf(x);
}

double SinCos(double x, double* pCos)
{
    *pCos = cos(x);
    return sin(x);
}
```

The resulting codegen for the helper is smaller https://godbolt.org/z/Ws7GPen8f on MSVC while it still uses the "faster" `sincos`. I would also expect some improvements in the codegen for managed portion of `Math.SinCos`.

**Comment by echesakovMSFT** [comment]
Chatted with Tanner offline and we both agreed that this will move to 7.0.0.

**Comment by conradakunga** [comment]
I've just run a benchmark on .NET 6 release 7 and (at least on my machine) it is orders of magnitude slower

```csharp
const double angle = 35;
[Benchmark]
public (double, double) GetIndividually()
{
    var sin = Math.Sin(angle);
    var cos = Math.Cos(angle);
    return (sin, cos);
}
 [Benchmark]
public (double, double) GetSimultaneously()
{
    var result = Math.SinCos(angle);
    return (result.Sin, result.Cos);
}
```

The results are as follows

```
BenchmarkDotNet=v0.13.1, OS=Windows 10.0.19043.1165 (21H1/May2021Update)
Intel Core i5-9300H CPU 2.40GHz, 1 CPU, 8 logical and 4 physical cores
.NET SDK=6.0.100-preview.7.21379.14
  [Host]     : .NET 6.0.0 (6.0.21.37719), X64 RyuJIT
  DefaultJob : .NET 6.0.0 (6.0.21.37719), X64 RyuJIT


|            Method |       Mean |     Error |    StdDev |
|------------------ |-----------:|----------:|----------:|
|   GetIndividually |  0.5548 ns | 0.0435 ns | 0.0806 ns |
| GetSimultaneously | 19.4477 ns | 0.3351 ns | 0.2616 ns |

```

Will this be the implementation that will be shipped in .NET 6?

**Comment by SingleAccretion** [comment]
The benchmark is not measuring what was intended - the Jit folds the expressions in question for the `GetIndividually` case.

**Comment by echesakovMSFT** [comment]
Un-assigning myself
cc @BruceForstall

**Comment by Anatoly-11** [comment]
Я вот пытался найти эту функцию и что-то не нашёл. Она есть в проце я это точно знаю. Но на каких процессорах?  Когда мы считаем sin с помощью команды fsincos мы одновременно 2 функции считаем за то же время.  Одна из них хранится на вершине сопроцессора st(0) и с помощью команды fpush мы извлекаем sin. А cos хранится в st(1) И его так же можно извлечь. Это в 2 раза быстрее чем по отдельности конечно. Что касается процессора ARM или других, то этого я сказать не могу... Может ли GPU делать такие вычисления? Полагаю что да. Ведь поворот это стандартное графическое преобразование вертексов и там надо обе функции одновременно.  Это то что должен уметь GPU. Но я не уверен. Может не все могут или что ещё.  Как это реализовано на языках высокого уровня типа С С++ C# не могу сказать. Потму тчо вопрос а как компилятор использует это или нет? И мю он когда мы пишем 2 таких функции ин понимает что надо вызывать не по одной а 2 сразу, т.е. оптимизирует.  У компиляторов есть некий ИИ и он возможно это делает за нас. и не надо ничего переписывать.  Но меня удивил факт тог очто я таки не нашёл эту функцию. Gро крайней мене у компилятора на msbuild.  Что уже странно. В gcc это есть... в clang тоже. Но вобщем то что msbuild далеко не up-to-date компилятор это известный факт.

**Comment by tannergooding** [comment]
Modern processors do not use the legacy x87 FPU pipeline as it is typically slow, inefficient, and can be inaccurate for some instructions.

Instead, most modern processors and operating systems (outside of some minor parameter passing conventions for 32-bit processes) exclusively use SSE or later instructions and use a polynomial based implementation to compute sin/cosine.

In the case of .NET, we currently defer to the underlying C runtime algorithm that exists in `libm` and this is simply a case where the transition from managed to native and difference in calling conventions is tripping up codegen for the call.

There are a few ways this can be improved and I plan on trying to improve that in .NET 9.

----------------

MSBuild is notably not the compiler, but rather a build system tool (more like `cmake` or `ninja`). Roslyn is the compiler for C# (more like `gcc` or `clang`) and RyuJIT is the compiler for IL (it's almost more similar to the `linker` in C/C++).
